ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionHeartbeatService.java

```diff
@@ -0,0 +1,256 @@
+    initialDelay = txnTimeout / 4;
+    period = txnTimeout / 2;
+  }
+
+  private static final class IMetaStoreClientFactory extends BasePooledObjectFactory<IMetaStoreClient> {
+
+    private final HiveConf conf;
+
+    @Override
+    public IMetaStoreClient create() throws Exception {
+      return HiveMetaStoreUtils.getHiveMetastoreClient(conf);
+    }
+
+    @Override
+    public PooledObject<IMetaStoreClient> wrap(IMetaStoreClient iMetaStoreClient) {
+      return new DefaultPooledObject<>(iMetaStoreClient);
+    }
+
+    @Override
+    public void destroyObject(PooledObject<IMetaStoreClient> p) {
+      p.getObject().close();
+    }
+
+    @Override
+    public boolean validateObject(PooledObject<IMetaStoreClient> p) {
+      //Not in use currently, would be good to validate the client at borrowing/returning, but this needs support from
+      //MetaStoreClient side
+      return super.validateObject(p);
+    }
+
+    public IMetaStoreClientFactory(HiveConf conf) {
+      this.conf = Objects.requireNonNull(conf);
+    }
+
+  }
+
+  private final class CompactionHeartbeater implements Runnable {
+    final private Logger LOG = LoggerFactory.getLogger(CompactionHeartbeater.class);
+    private final long txnId;
+    private final long lockId;
+    private final String tableName;
+    private final Object lock = new Object();
+    private volatile boolean running = false;
+
+
+    @Override
+    public void run() {
+      IMetaStoreClient msc = null;
+      try {
+        synchronized (lock) {
+          running = true;
+        }
+        LOG.debug("Heartbeating compaction transaction id {} for table: {}", txnId, tableName);
+        // Create a metastore client for each thread since it is not thread safe
+        msc = clientPool.borrowObject();
+        msc.heartbeat(txnId, lockId);
+        clientPool.returnObject(msc);
+      } catch (NoSuchElementException nsee) {
+        LOG.error("Compaction transaction heartbeater pool exhausted, unable to heartbeat", nsee);
+      } catch (Exception e) {
+        LOG.error("Error while heartbeating compaction transaction id {} for table: {}", txnId, tableName, e);
+        try {
+          clientPool.invalidateObject(msc);
+        } catch (Exception ex) {
+          LOG.error("Error while invalidating a broken MetaStoreClient instance", e);
+        }
+      } finally {
+        synchronized (lock) {
+          running = false;
+          lock.notifyAll();
+        }
+      }
+    }
+
+    public void waitUntilFinish() throws InterruptedException {
+      synchronized (lock) {
+        if (running) {
+          lock.wait();
+        }
+      }
+    }
+
+    private CompactionHeartbeater(long txnId, long lockId, String tableName) {
+      this.tableName = Objects.requireNonNull(tableName);
+      this.txnId = txnId;
+      this.lockId = lockId;
+    }
+
+  }
+
+  private static final class TaskWrapper {
```

Comment:
why do we need this?

Reply:
It ties CompactionHeartbeater and the returned future together. If you check the `stopHeartbeat()` method, the future is required for cancelling the task in the executor, and the heartbeater instance is required to call the `waitUntilFinish()` method, which will wait for finishing the hearbeat if that's currently executing. This is required to avoid heartbeats after txn abort or commit.

Reply:
see above